core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java

```diff
@@ -1871,6 +1910,76 @@ private ExprValue exprTimestampAddForTimeType(Clock clock,
     return exprTimestampAdd(partExpr, amountExpr, new ExprDatetimeValue(datetime));
   }
 
+  private ExprValue getTimeDifference(String part, LocalDateTime startTime, LocalDateTime endTime) {
+    long returnVal;
+    switch (part) {
+      case "MICROSECOND":
+        returnVal = MICROS.between(startTime, endTime);
+        break;
+      case "SECOND":
+        returnVal = SECONDS.between(startTime, endTime);
+        break;
+      case "MINUTE":
+        returnVal = MINUTES.between(startTime, endTime);
+        break;
+      case "HOUR":
+        returnVal = HOURS.between(startTime, endTime);
+        break;
+      case "DAY":
+        returnVal = DAYS.between(startTime, endTime);
+        break;
+      case "WEEK":
+        returnVal = WEEKS.between(startTime, endTime);
+        break;
+      case "MONTH":
+        returnVal = MONTHS.between(startTime, endTime);
+        break;
+      case "QUARTER":
+        returnVal = MONTHS.between(startTime, endTime) / 3;
+        break;
+      case "YEAR":
+        returnVal = YEARS.between(startTime, endTime);
+        break;
+      default:
+        return ExprNullValue.of();
+    }
+    return new ExprLongValue(returnVal);
+  }
+
+  private ExprValue exprTimestampDiff(
+                                      ExprValue partExpr,
+                                      ExprValue startTimeExpr,
+                                      ExprValue endTimeExpr) {
+    return getTimeDifference(
+        partExpr.stringValue(),
+        startTimeExpr.datetimeValue(),
+        endTimeExpr.datetimeValue());
+  }
+
+  private ExprValue exprTimestampDiffForTimeType(Clock queryStart,
+                                                 ExprValue partExpr,
+                                                 ExprValue startTimeExpr,
+                                                 ExprValue endTimeExpr) {
+    String part = partExpr.stringValue();
+    LocalDateTime startTime;
+    LocalDateTime endTime;
+
+    //Fill in date parts if a time argument is provided
+    if (startTimeExpr instanceof ExprTimeValue) {
+      startTime = LocalDateTime.of(LocalDate.now(queryStart), startTimeExpr.timeValue());
+    } else {
+      startTime = startTimeExpr.datetimeValue();
+    }
+
+    if (endTimeExpr instanceof ExprTimeValue) {
+      endTime = LocalDateTime.of(LocalDate.now(queryStart), endTimeExpr.timeValue());
+    } else {
+      endTime = endTimeExpr.datetimeValue();
+    }
```

Comment:
Pass `FunctionProperties` as a first argument instead of `Clock` and you can simplify your function by using `DateTimeUtils.extractDateTime`:
```java
  return getTimeDifference(partExpr.stringValue(), extractDateTime(startTimeExpr, fp), extractDateTime(endTimeExpr, fp));
```

Reply:
Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/251/commits/ed97a24f7ab7d040513331c17586b72bebdf6eb1